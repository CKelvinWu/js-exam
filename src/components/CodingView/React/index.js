import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';
import 'brace';
import 'brace/mode/jsx';
import 'brace/mode/javascript';
import 'brace/theme/textmate';
import 'brace/theme/monokai';

import Grid from 'components/Grid';
import GridItem from 'components/Grid/GridItem';
import ConsoleWidget from 'components/Widgets/ConsoleWidget';
import CodeWidget from 'components/Widgets/CodeWidget';
import ResultWidget from 'components/Widgets/ResultWidget';
import AnswerWidget from 'components/Widgets/AnswerWidget';

import debouncedRunCode from 'utils/runCode';
import { REACT as GRID_LABEL_REACT } from 'utils/gridLabel';

import styles from './ReactPage.module.scss';

class ReactPage extends Component {
  static propTypes = {
    isExaming: PropTypes.bool,
    code: PropTypes.string,
    compiledCode: PropTypes.string,
    consoleMsg: PropTypes.array,
    handleCodeChange: PropTypes.func,
    wrappedConsole: PropTypes.object,
    resetConsole: PropTypes.func,
  };

  controlHeight = 70;

  componentDidMount() {
    const { compiledCode, wrappedConsole, resetConsole } = this.props;
    resetConsole();
    debouncedRunCode({ code: compiledCode, wrappedConsole });
  }

  shouldComponentUpdate(nextProps) {
    const { compiledCode: previousCompiledCode } = this.props;
    const { compiledCode, wrappedConsole, resetConsole } = nextProps;
    if (previousCompiledCode !== compiledCode) {
      resetConsole();
      // WARNING: This is not debounced
      debouncedRunCode({ code: compiledCode, wrappedConsole });
    }
    return true;
  }

  render() {
    const { isExaming, code, handleCodeChange, consoleMsg } = this.props;
    const layout = [
      {
        key: 'code',
        x: 0,
        y: 0,
        width: window.innerWidth / 2 + 50,
        height: window.innerHeight / 2 + 100,
        minWidth: 100,
        minHeight: 100,
        maxWidth: 700,
        maxHeight: 500,
      },
      {
        key: 'answer',
        x: 0,
        y: 1,
        width: window.innerWidth / 2 + 50,
        height: window.innerHeight / 2 - 150,
        minWidth: 100,
        maxWidth: 700,
      },
      {
        key: 'result',
        x: 1,
        y: 0,
        width: window.innerWidth / 2 - 50,
        height: (window.innerHeight - this.controlHeight) / 2,
        minWidth: 100,
        minHeight: 100,
        maxWidth: 700,
        maxHeight: 500,
      },
      {
        key: 'console',
        x: 1,
        y: 1,
        width: window.innerWidth / 2 - 50,
        height: (window.innerHeight - this.controlHeight) / 2 - 50,
        minWidth: 100,
        minHeight: 100,
        maxWidth: 700,
        maxHeight: 500,
      },
    ];
    return (
      <div className={styles.app}>
        <Grid layout={layout} totalWidth="100%" totalHeight="100%" autoResize>
          <GridItem key="code" label={GRID_LABEL_REACT.code}>
            <CodeWidget
              handleCodeChange={handleCodeChange}
              data={code}
              readOnly={!isExaming}
              mode="jsx"
              theme="monokai"
            />
          </GridItem>
          <GridItem key="answer" label={GRID_LABEL_REACT.answer}>
            {isExaming ? <AnswerWidget /> : null}
          </GridItem>
          <GridItem key="result" label={GRID_LABEL_REACT.result}>
            {isExaming ? <ResultWidget /> : null}
          </GridItem>
          <GridItem key="console" label={GRID_LABEL_REACT.console}>
            {isExaming ? <ConsoleWidget data={consoleMsg} /> : null}
          </GridItem>
        </Grid>
      </div>
    );
  }
}

export default withRouter(ReactPage);
