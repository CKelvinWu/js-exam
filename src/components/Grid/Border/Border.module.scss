.border {
  width: 100%;
  height: 100%;
  background: #e0e0e0;
}

.border > div {
  border-right: 1px solid #e0e0e0;
}

.vertical {
  cursor: row-resize;
}

.horizontal {
  cursor: col-resize;
}
