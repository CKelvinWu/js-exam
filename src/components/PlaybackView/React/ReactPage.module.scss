.app {
  display: flex;
  flex-direction: row;
}

.rightWidget {
  height: 33.3%;
  width: 100%;
  border: 1px solid #eee;
}

.label {
	position: absolute;
  margin-top: 5px;
  right: 17px;
	z-index: 10;
	padding: 2px 12px;
	color: white;
	background-color: rgba($color: #000000, $alpha: 0.5);
}