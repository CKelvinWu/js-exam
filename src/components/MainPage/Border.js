import React, { Component } from 'react';
import './MainPage.css';

class Border extends Component {
  constructor(props) {
    super(props);
    this.ref = React.createRef();
    this.resize = this.resize.bind(this);
    this.startResize = this.startResize.bind(this);
    this.endResize = this.endResize.bind(this);
    this.state = { width: null };
  }
  componentDidMount() {
    this.setState({ width: this.ref.current.clientWidth });
  }
  startResize(e) {
    if (e.target.contains(this.ref.current)) {
      window.addEventListener('mousemove', this.resize);
      window.addEventListener('mouseup', this.endResize);
    }
  }
  resize(e) {
    this.setState({ width: e.clientX - this.ref.current.offsetLeft });
  }
  endResize(e) {
    window.removeEventListener('mousemove', this.resize);
    window.removeEventListener('mouseup', this.endResize);
  }
  render() {
    return (
      <div
        ref={this.ref}
        className={`border ${this.props.className ? this.props.className : ''}`}
        onMouseDown={this.startResize}
        style={
          this.state.width === null ?
            {} :
            { width: this.state.width + 'px' }
        }
      >
        { this.props.children }
      </div>
    )
  }
}

export default Border;