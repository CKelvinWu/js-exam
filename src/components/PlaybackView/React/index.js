import React, { Component } from 'react';
import PropTypes from 'prop-types';
import 'brace';
import 'brace/mode/jsx';
import 'brace/mode/javascript';
import 'brace/theme/textmate';
import 'brace/theme/monokai';

import { Spin } from 'antd';

import CodeWidget from 'components/Widgets/CodeWidget';
import ResultWidget from 'components/Widgets/ResultWidget';
import AnswerWidget from 'components/Widgets/AnswerWidget';

import debouncedRunCode from 'utils/runCode';
import { REACT as GRID_LABEL_REACT } from 'utils/gridLabel';

import styles from './ReactPage.module.scss';

class ReactPage extends Component {
  constructor(props) {
    super(props);
    this.state = {
      height: props.height,
      widthCode: props.width,
      widthTests: props.width,
    };
  }

  async componentDidMount() {
    const { compiledCode, addTape } = this.props;

    this.setState({
      height: `${window.innerHeight - 160}px`,
      widthCode: `${window.innerWidth * 0.45}px`,
      widthTests: `${window.innerWidth * 0.3}px`,
    });
    debouncedRunCode({ code: compiledCode, onTapeUpdate: addTape });
  }

  shouldComponentUpdate(nextProps) {
    
    const { compiledCode: previousCompiledCode } = this.props;
    const { compiledCode } = nextProps;
    if (previousCompiledCode !== compiledCode) {
      debouncedRunCode({ code: compiledCode });
    }
    return true;
  }

  render() {
    this.codeWidgetstyle = {
      height: this.state.height,
      width: this.state.widthCode,
      position: 'relative',
    };
    this.testWidgetStyle = {
      height: this.state.height,
      width: this.state.widthTests,
      position: 'relative',
    };
    const { handleCodeChange, test, code } = this.props;
    return (
      <React.Fragment>
        <div className={styles.app}>
          <div style={this.codeWidgetstyle}>
            <div className={styles.label}>Code</div>
            <CodeWidget
              handleCodeChange={handleCodeChange}
              data={code}
              mode="jsx"
              theme="monokai"
              readOnly
            />
          </div>
          <div style={this.testWidgetStyle}>
            <div className={styles.rightWidget}>
              <div className={styles.label}>{GRID_LABEL_REACT.test}</div>
              <CodeWidget data={test} mode="jsx" theme="textmate" readOnly />
            </div>
            <div className={styles.rightWidget}>
              <div className={styles.label}>{GRID_LABEL_REACT.answer}</div>
              <AnswerWidget />
            </div>
            <div className={styles.rightWidget}>
              <div className={styles.label}>{GRID_LABEL_REACT.result}</div>
              <ResultWidget />
            </div>
          </div>
        </div>
      </React.Fragment>
    );
  }
}

ReactPage.propTypes = {
  compiledCode: PropTypes.string,
  code: PropTypes.string,
  height: PropTypes.string,
  width: PropTypes.string,
  test: PropTypes.string,
  handleCodeChange: PropTypes.func,
  addTape: PropTypes.func,
};

export default ReactPage;
