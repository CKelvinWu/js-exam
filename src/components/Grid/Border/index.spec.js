import React from 'react';
import { mount } from 'enzyme';
import Border from './index';

describe('Border index spec', () => {
  it('should render border component properly', () => {
    const element = mount(
      <Border
        allowWidth
        allowHeight
        borderSize={1}
        width={1}
        height={10}
        maxHeight={10}
        minHeight={1}
        onUpdate={jest.fn()}
        disabled={false}
      />,
    );
    expect(element.render()).toMatchSnapshot();
  });
});
