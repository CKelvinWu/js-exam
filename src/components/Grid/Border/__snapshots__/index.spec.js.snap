// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Border index spec should render border component properly 1`] = `
<div
  class="border  horizontal vertical"
  style="width: 1px; height: 10px; padding-right: 1px; padding-bottom: 1px;"
/>
`;
